crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -77,12 +77,158 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
     duration.as_nanos().try_into().map_err(Into::into)
 }
 
+fn stdin_nonempty() -> bool {
+    use std::io::Read;
+    std::io::stdin().bytes().peekable().peek().is_some()
+}
+
 pub(crate) fn poll_oneoff(
     timeout: Option<ClockEventData>,
     fd_events: Vec<FdEventData>,
     events: &mut Vec<wasi::__wasi_event_t>,
-) -> Result<Vec<wasi::__wasi_event_t>> {
-    unimplemented!("poll_oneoff")
+) -> Result<()> {
+    use crate::fdentry::Descriptor;
+    if fd_events.is_empty() && timeout.is_none() {
+        return Ok(());
+    }
+
+    // Currently WASI file support is only (a) regular files (b) directories (c) symlinks on Windows,
+    // which are always ready to write on Unix.
+    //
+    // We need to consider stdin/stdout/stderr separately.
+    // We treat stdout/stderr as always ready to write. I'm not sure if it's correct
+    // on Windows but I have not find any way of checking if a write to stdout would block.
+    // Therefore, we only poll the stdin.
+    let mut stdin_events = vec![];
+    let mut immediate_events = vec![];
+
+    for event in fd_events {
+        match event.descriptor {
+            Descriptor::Stdin if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ => {
+                stdin_events.push(event)
+            }
+            _ => immediate_events.push(event),
+        }
+    }
+
+    // we have at least one immediate event, so we don't need to care about stdin
+    if immediate_events.len() > 0 {
+        for event in immediate_events {
+            let size = match event.descriptor {
+                Descriptor::OsFile(os_file) if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ => {
+                    os_file
+                        .metadata()
+                        .expect("FIXME return a proper error")
+                        .len()
+                }
+                Descriptor::Stdin => panic!("Descriptor::Stdin should have been filtered out"),
```

Comment:
Is it possible to hit this panic by requesting a write event for stdin?

Reply:
I guess it doesn't hurt to leave that in. Now, indeed I cannot see it ever triggering, but in the future, when we touch this bit of code, it might actually come in handy. To make our intent about this one clearer, we could swap `panic` for `unreachable`.

Reply:
I consider easily reachable panics from WASI to be bugs as we should only do so for  violations of invariant conditions to catch implementation bugs.  The user passing in invalid arguments must result in a error code return value and not a panic.

Reply:
Oh sorry, I misread this. If by this point it is still possible to have a situation where we request a (`stdin`, `event.r#type == wasi::__WASI_EVENTTYPE_FD_WRITE`), then indeed this event will end up in `immediate_events` tripping at the panic. However, I though that `stdin` doesn't come with `FD_WRITE` rights hence this type of situation getting filtered out in the common code between hosts [here](https://github.com/bytecodealliance/wasmtime/blob/master/crates/wasi-common/src/hostcalls_impl/misc.rs#L250). I might be wrong of course, in which case, we should definitely throw an actual error here like you suggested.

Reply:
If we are treating it already as an access error then I agree we can just resolve this comment as one of those invariant assertions that is completely valid.

Reply:
We'd also have to check if we're doing the same for stdout/stderr too (are we erroring in the common code already for a read? this implementation it to return 0 bytes available).

Reply:
This code doesn't exist anymore, resolving.